File: formulus/android/app/src/main/java/org/opendataensemble/formulus/ProfileDatabaseLifecycle.java
package org.opendataensemble.formulus;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.regex.Pattern;

/** Filesystem-only lifecycle guard. Every adapter must be prepared before it is constructed. */
final class ProfileDatabaseLifecycle {
    private static final Pattern DATABASE_NAME = Pattern.compile(
        "formulus(?:_[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})?"
    );
    private static final String[] SUFFIXES = {".db", ".db-journal", ".db-wal", ".db-shm"};
    // Never clear these during module invalidation or ReactHost reload. Only process death resets them.
    private static final Set<String> OPENED_DATABASES = new HashSet<>();
    private static final Set<String> RETIRED_DATABASES = new HashSet<>();

    private final File directory;

    ProfileDatabaseLifecycle(File directory) {
        this.directory = directory;
    }

    static String generateProfileId() {
        return UUID.randomUUID().toString();
    }

    private static void validateName(String name) {
        if (name == null || !DATABASE_NAME.matcher(name).matches()) {
            throw new IllegalArgumentException("Expected formulus or formulus_<lowercase UUID>");
        }
    }

    void prepareDatabase(String name) {
        synchronized (ProfileDatabaseLifecycle.class) {
            validateName(name);
            if (RETIRED_DATABASES.contains(name)) {
                throw new IllegalStateException("A database scheduled for deletion cannot be reopened");
            }
            // Retain every prepared name across module and runtime replacement.
            OPENED_DATABASES.add(name);
        }
    }

    boolean isDatabaseOpen(String name) {
        synchronized (ProfileDatabaseLifecycle.class) {
            validateName(name);
            // Conservative 'ever prepared', not a probe of SQLite's actual connection state.
            return OPENED_DATABASES.contains(name);
        }
    }

    boolean databaseExists(String name) throws IOException {
        synchronized (ProfileDatabaseLifecycle.class) {
            validateName(name);
            // Inspect directory entries, never a database descriptor. Count even dangling
            // symlinks/non-regular entries as legacy evidence rather than a fresh install.
            Set<String> entries = listDirectory(directory.getCanonicalFile());
            for (String suffix : SUFFIXES) {
                if (entries.contains(name + suffix)) {
                    return true;
                }
            }
            return false;
        }
    }

    boolean deleteDatabase(String name) throws IOException {
        synchronized (ProfileDatabaseLifecycle.class) {
            validateName(name);
            if (OPENED_DATABASES.contains(name)) {
                return false;
            }
            // Prevent a prepare racing with a failed/partial deletion. The registry must also
            // retain its durable tombstone and never reuse a deleted name, including after death.
            RETIRED_DATABASES.add(name);
            File root = directory.getCanonicalFile();
            Set<String> entries = listDirectory(root);
            File[] files = new File[SUFFIXES.length];
            for (int i = 0; i < SUFFIXES.length; i++) {
                File file = new File(root, name + SUFFIXES[i]);
                files[i] = file;
                if (entries.contains(file.getName()) &&
                    (!file.getCanonicalFile().equals(file) || !file.isFile())) {
                    throw new IOException("Refusing to delete a non-regular database file: " + file.getName());
                }
            }
            // Remove the main file first. On any failure, leave the tombstone for an idempotent retry.
            for (File file : files) {
                if (entries.contains(file.getName()) && !file.delete()) {
                    throw new IOException("Could not delete database file: " + file.getName());
                }
            }
            Set<String> remaining = listDirectory(root);
            for (File file : files) {
                if (remaining.contains(file.getName())) {
                    throw new IOException("Database file still exists: " + file.getName());
                }
            }
            return true;
        }
    }

    private static Set<String> listDirectory(File root) throws IOException {
        // File.exists() alone cannot distinguish absence from an I/O/permission error.
        // File APIs (rather than java.nio.file) also support Formulus's Android API 24 minimum.
        String[] names = root.list();
        if (names == null) {
            throw new IOException("Could not inspect the WatermelonDB directory");
        }
        return new HashSet<>(Arrays.asList(names));
    }
}

File: formulus/ios/Formulus/ProfileDatabaseLifecycle.m
#import "ProfileDatabaseLifecycle.h"
#import <errno.h>
#import <sys/stat.h>
#import <unistd.h>

// Never reset on native module invalidation or RN reload. Only process death clears these.
static NSMutableSet<NSString *> *openedDatabases;
static NSMutableSet<NSString *> *retiredDatabases;
static NSRegularExpression *databaseNamePattern;

static BOOL Fail(NSError **error, ProfileDatabaseError code, NSString *message) {
  if (error) {
    *error = [NSError errorWithDomain:@"FormulusProfileDatabase" code:code
                            userInfo:@{NSLocalizedDescriptionKey: message}];
  }
  return NO;
}

static BOOL ValidateName(NSString *name, NSError **error) {
  if (![name isKindOfClass:NSString.class] ||
      ![databaseNamePattern firstMatchInString:name options:0 range:NSMakeRange(0, name.length)]) {
    return Fail(error, ProfileDatabaseErrorName, @"Expected formulus or formulus_<lowercase UUID>");
  }
  return YES;
}

@implementation ProfileDatabaseLifecycle {
  NSString *_directory;
}

+ (void)initialize {
  if (self == ProfileDatabaseLifecycle.class) {
    openedDatabases = [NSMutableSet new];
    retiredDatabases = [NSMutableSet new];
    databaseNamePattern = [NSRegularExpression regularExpressionWithPattern:
      @"\\Aformulus(?:_[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})?\\z"
      options:0 error:NULL];
  }
}

+ (NSString *)generateProfileId {
  return NSUUID.UUID.UUIDString.lowercaseString;
}

- (instancetype)initWithDirectory:(NSString *)directory {
  if ((self = [super init])) {
    _directory = [directory copy];
  }
  return self;
}

- (BOOL)prepareDatabase:(NSString *)name error:(NSError **)error {
  @synchronized (ProfileDatabaseLifecycle.class) {
    if (!ValidateName(name, error)) {
      return NO;
    }
    if ([retiredDatabases containsObject:name]) {
      return Fail(error, ProfileDatabaseErrorState, @"A database scheduled for deletion cannot be reopened");
    }
    // Retain every prepared name across module and runtime replacement.
    [openedDatabases addObject:name];
    return YES;
  }
}

- (NSNumber *)isDatabaseOpen:(NSString *)name error:(NSError **)error {
  @synchronized (ProfileDatabaseLifecycle.class) {
    if (!ValidateName(name, error)) {
      return nil;
    }
    return @([openedDatabases containsObject:name]);
  }
}

- (NSNumber *)databaseExists:(NSString *)name error:(NSError **)error {
  @synchronized (ProfileDatabaseLifecycle.class) {
    if (!ValidateName(name, error)) {
      return nil;
    }
    // Listing names never opens a database descriptor. Include dangling symlinks
    // and non-regular entries as evidence, rather than silently treating them as absent.
    NSArray<NSString *> *entries = [NSFileManager.defaultManager
      contentsOfDirectoryAtPath:_directory error:NULL];
    if (!entries) {
      Fail(error, ProfileDatabaseErrorIO, @"Could not inspect the WatermelonDB directory");
      return nil;
    }
    for (NSString *suffix in @[@".db", @".db-journal", @".db-wal", @".db-shm"]) {
      if ([entries containsObject:[name stringByAppendingString:suffix]]) {
        return @YES;
      }
    }
    return @NO;
  }
}

- (NSNumber *)deleteDatabase:(NSString *)name error:(NSError **)error {
  @synchronized (ProfileDatabaseLifecycle.class) {
    if (!ValidateName(name, error)) {
      return nil;
    }
    if ([openedDatabases containsObject:name]) {
      return @NO;
    }
    // Block reopening after a partial deletion. The registry must retain its durable
    // tombstone across cold launches, and must never reuse deleted names.
    [retiredDatabases addObject:name];
    NSString *root = [_directory stringByResolvingSymlinksInPath];
    struct stat info;
    if (stat(root.fileSystemRepresentation, &info) != 0 || !S_ISDIR(info.st_mode)) {
      Fail(error, ProfileDatabaseErrorIO, @"Could not inspect the WatermelonDB directory");
      return nil;
    }
    NSMutableArray<NSString *> *paths = [NSMutableArray new];
    for (NSString *suffix in @[@".db", @".db-journal", @".db-wal", @".db-shm"]) {
      NSString *path = [root stringByAppendingPathComponent:[name stringByAppendingString:suffix]];
      [paths addObject:path];
      if (lstat(path.fileSystemRepresentation, &info) == 0) {
        if (!S_ISREG(info.st_mode)) {
          Fail(error, ProfileDatabaseErrorIO,
               [@"Refusing to delete a non-regular database file: " stringByAppendingString:path.lastPathComponent]);
          return nil;
        }
      } else if (errno != ENOENT) {
        Fail(error, ProfileDatabaseErrorIO,
             [@"Could not inspect database file: " stringByAppendingString:path.lastPathComponent]);
        return nil;
      }
    }
    // No SQLite engine is opened, no file contents read, no recursive directory removal.
    // Main file first; failures retain the tombstone for an idempotent retry.
    for (NSString *path in paths) {
      if (unlink(path.fileSystemRepresentation) != 0 && errno != ENOENT) {
        Fail(error, ProfileDatabaseErrorIO,
             [@"Could not delete database file: " stringByAppendingString:path.lastPathComponent]);
        return nil;
      }
    }
    for (NSString *path in paths) {
      if (lstat(path.fileSystemRepresentation, &info) == 0 || errno != ENOENT) {
        Fail(error, ProfileDatabaseErrorIO,
             [@"Database file is not confirmed absent: " stringByAppendingString:path.lastPathComponent]);
        return nil;
      }
    }
    return @YES;
  }
}
@end
